src/main/vcp_hal/usbd_cdc_interface.c

```diff
@@ -379,16 +397,15 @@ uint32_t CDC_Send_FreeBytes(void)
  */
 uint32_t CDC_Send_DATA(const uint8_t *ptrBuffer, uint32_t sendLength)
 {
-    USBD_CDC_HandleTypeDef *hcdc = (USBD_CDC_HandleTypeDef*)USBD_Device.pCDC_ClassData;
-    while (hcdc->TxState != 0);
-
-    for (uint32_t i = 0; i < sendLength; i++)
-    {
-        UserTxBuffer[UserTxBufPtrIn] = ptrBuffer[i];
-        UserTxBufPtrIn = (UserTxBufPtrIn + 1) % APP_TX_DATA_SIZE;
+    for (uint32_t i = 0; i < sendLength; i++) {
         while (CDC_Send_FreeBytes() == 0) {
+            // block until there is free space in the ring buffer
             delay(1);
         }
+        ATOMIC_BLOCK(NVIC_BUILD_PRIORITY(6, 0)) { // Paranoia
```

Comment:
It would be cleaner and faster to copy data into buffer outside atomic, using local variable for UserTxBufPtrIn and then atomically set UserTxBufPtrIn.
Also this code can be replaced by 2 `memcpy`